### PR TITLE
Block device support for domains

### PR DESCRIPTION
**Pull request by muroj**
My first pass at providing support for block devices. Attempts to address #662 and #418. (Disclaimer: I am new to go).

### CODE REVIEW

File: libvirt/resource_libvirt_volume.go [hunk 1]
```diff
@@ -254,10 +254,10 @@ func resourceLibvirtVolumeCreate(d *schema.ResourceData, meta interface{}) error
 	if _, ok := d.GetOk("source"); ok {
 		err = img.Import(newCopier(client.libvirt, volume, volumeDef.Capacity.Value), volumeDef)
 		if err != nil {
-		//  don't save volume ID  in case of error. This will taint the volume after.
-		// If we don't throw away the id, we will keep instead a broken volume.
-		// see for reference: https://github.com/dmacvicar/terraform-provider-libvirt/issues/494
-		d.Set("id", "")	
+			//  don't save volume ID  in case of error. This will taint the volume after.
```
Comment (MalloZup) [inline libvirt/resource_libvirt_volume.go]: perhaps we can remove the identation here since not related but ok

**Comment by MalloZup** [comment]
checkout travis is failing

```
+go test -v -covermode=count -coverprofile=profile.cov -timeout=1200s -test.failfast ./libvirt

# github.com/dmacvicar/terraform-provider-libvirt/libvirt [github.com/dmacvicar/terraform-provider-libvirt/libvirt.test]

libvirt/resource_libvirt_domain_test.go:1221:43: cmd.String undefined (type *exec.Cmd has no field or method String)

libvirt/resource_libvirt_domain_test.go:1240:43: cmd.String undefined (type *exec.Cmd has no field or method String)

FAIL	github.com/dmacvicar/terraform-provider-libvirt/libvirt [build failed]

Makefile:36: recipe for target 'testacc' failed

make: *** [testacc] Error 2```

**Comment by muroj** [comment]
I am running same build commands locally and not seeing the errors:

```
:~/go/src/github.com/muroj/terraform-provider-libvirt$ make build
go fmt ./libvirt .
go run golang.org/x/lint/golint -set_exit_status ./libvirt .
go vet ./libvirt
```

Is it because I am running `go1.13.4` instead of `go1.11.13`?

**Comment by MalloZup** [comment]
@muroj  dunno I didn't tried but try it :grin:

File: libvirt/resource_libvirt_volume.go (hunk 1 above)
Comment (muroj) [reply libvirt/resource_libvirt_volume.go]: I tried reverting it back to original. I think the changes were the result of `go-fmt`.

**Comment by MalloZup** [comment]
@muroj thx for pr great job

**Comment by dmacvicar** [comment]
While trying to do a release I realize this PR broke integration tests.

It requires both root access to test this, and on top, it assumes _libvirtd_ is running on the same machine as the test, and assumes _/usr/sbin_ is on _$PATH_ (eg.  mkfs.ext4).

**Comment by dmacvicar** [comment]
Managed to fix this calling executables with full path (sbin). I am still stuck in:

```
Jan 03 13:36:32 piscolita.suse.de libvirtd[5524]: internal error: process exited while connecting to monitor: 2020-01-03T12:36:32.234735Z qemu-system-x86_64: -blockdev {"driver":"host_device","filename":"/tmp/cgvuajorbpkmtkbuetximqhkpcetcwlby","node-name":"libvirt-1-storage","auto-read-only":true,"discard":"unmap"}: 'host_device' driver expects either a character or block device
```

Which I guess comes from:

```golang
var configBlockDevice = fmt.Sprintf(`

resource "libvirt_domain" "%s" {
	name = "%s"
	
	disk {
		block_device = "%s"
	}

}`, randomDomainName, randomDomainName, tmpfile)
```

Which makes total sense. `tmpfile` is not a device. The loop device is a device. I wonder how this works in the libvirt used in travis.

If I change the code to use the loop device, I fall into this:

```golang
Error: Error creating libvirt domain: virError(Code=1, Domain=10, Message='internal error: process exited while connecting to monitor: 2020-01-03T13:19:32.338303Z qemu-system-x86_64: -blockdev {"driver":"host_device","filename":"/dev/loop0","node-name":"libvirt-1-storage","auto-read-only":true,"discard":"unmap"}: The device is not writable: Permission denied')
```

Which makes sense, as _losetup_ is called as:

```golang
// Mount the file to a loop device
cmd = exec.Command("sudo", "/sbin/losetup", "--read-only", strings.TrimRight(string(loopdev), "\n"), filename)
fmt.Printf("Executing command: %s\n", strings.Join(cmd.Args, " "))
```

Which means we are setting up a _rw_ disk on a _ro_ device. However, we haven't implemented _readonly_ for disks.

Fixing _losetup_, I need to fix the check function to check that the source is the device.

Fixing that, I get an error that _tmpfile_ is not used. Looks like the temporary file is not cleaned. I have fixed that as well.

Fixes being prepared here:
https://github.com/dmacvicar/terraform-provider-libvirt/tree/dmacvicar_fix_tests_locally

**Comment by MalloZup** [comment]
@dmacvicar thx and sorry for not having catching it up.

Let me know if I can help somehow. Thx!

**Comment by dmacvicar** [comment]
I put all the fixes in master before doing the release! thanks!

The PR was passing testacc fine on Travis! so not a review problem. openSUSE/libvirt were slightly different.

**Comment by muroj** [comment]
@dmacvicar @MalloZup  Thanks both for your work (and patience) with me on this.

**Pull request merged**